embedded-workshop-book/src/uarte-implementation.md

```diff
@@ -0,0 +1,183 @@
+
+<details>
+  <summary>Solution</summary>
+
+```rust
+  let pins =  hal::uarte::Pins {
+            rxd: pins.p0_08.degrade().into_floating_input(),
+            txd: pins.p0_06.degrade().into_push_pull_output(Level::High),
+            cts: Some(pins.p0_07.degrade().into_floating_input()),
+            rts: Some(pins.p0_05.degrade().into_push_pull_output(Level::High)),
+        };
+       
+
+        let uarte = hal::uarte::Uarte::new(periph.UARTE1, pins, Parity::INCLUDED, Baudrate::BAUD115200);
+```
+</details>
+
+### Step 5: Board struct
+
+✅  Add a field for the `Uarte` struct in the `Board` struct. 
+add the field to the instance of the `Board` struct in `fn init()`.
+
+<details>
+  <summary>Solution</summary>
+
+```rust
+
+pub struct Board {
+    /// LEDs
+    pub leds: Leds,
+    /// Buttons
+    pub buttons: Buttons,
+    /// Timer
+    pub timer: Timer,
+    /// uarte interface
+    pub uarte: Uarte,
+}
+
+// ...
+
+pub fn init() -> Result<Board, ()> {
+
+    // ... 
+
+        Ok(Board {
+            leds: Leds {
+                // ...
+            },
+
+            buttons: Buttons {
+                // ...
+            },
+            // 🔼  --- Exercise Button --- 🔼 
+
+            timer: Timer { inner: timer },
+
+            uarte: Uarte { inner: uarte },
+        })
+    } else {
+        Err(())
+    }
+```
+
+</details>
+
+### Step 6: Implementing the `fmt::Write` trait
+
+We can't just write to the `Uarte` instance. A simple write would write from flash memory. This does not work because of EasyDMA. We have to write a function that implements the `fmt::Write` trait. This trait guarantees that the buffer is fully and successfully written on a stack allocated buffer, before it returns. 
+
+✅ Add `use::core::fmt;` to your imports.
+
+✅ Create a public method `write_str`. It takes a mutable reference to self and a `&str` as argument. It returns an `fmt::Result`
+
+✅ Create a buffer. The type is an `array` of 16 u8, set to all 0. 
+
+✅ To copy all data into an on-stack buffer, iterate over every chunk of the string to copy it into the buffer.
+
+<details>
+  <summary>Solution</summary>
+
+```rust
+impl fmt::Write for Uarte {
+
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        // Copy all data into an on-stack buffer so we never try to EasyDMA from
+        // flash.
+        let mut buf: [u8; 16] = [0; 16];
+        for block in s.as_bytes().chunks(16) {
+            buf[..block.len()].copy_from_slice(block);
+            self.inner.write(&buf[..block.len()]).map_err(|_| fmt::Error)?;
+        }
+
+        Ok(())
+    }
+}
+```
+</details>
+
+### Step 7: Connect your computer to the virtual UART
+[todo!] [directions for mac present, linux and windows are missing.]
+   
+✅  Use the following command to find the address of the nRF52840-DK on your computer. 
+
+```
+ls /dev/tty.usbmodem*
+```
+
+✅  Run the following command to run `screen` with the nRF52840-DK with 115200 baud. 
```

Comment:
`screen` is a bit of a pain to exit and doesn't have a great UI. I have had a lot of success with `python3 -m serial.tools.miniterm` (or the alias `pyserial-miniterm`), but that requires you have Python and pyserial installed. PuTTY and TeraTerm are other popular options (especially on Windows).